---
title: "class 18"
author: "DongSu Kim"
date: "3/5/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Connect to the NCI-GDC database
The NCI-GDC (cancer genomics commons) is a major portal for cancer genomics data.

Here we use the bioconductor package **GenomicDataCommons** to talk to this database and download data etc. all from R.

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

Can we talk to the online database?
```{r}
status()
```


## Querying the GDC from R

The pipe operator can be useful in R and sometimes is more clear to read. 
```{r}
c(1:10)%>% mean()
```


```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

### Plot thos overview data 

```{r}
x <- cases_by_project$project.project_id

barplot(x$doc_count, names.arg = x$key, las=2, col="lightblue", log="y" )
```


## Variant analysis with R

Find files available for PAAD project that we can then examine variants from

```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
maf.file
```


Read in the MAF file we found
```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)
```

Now make a summary plot of this data
```{r}
plotmafSummary(vars)
```


## Oncoplots

We can make oncoplots for a specific gene too...
```{r}
lollipopPlot(vars, gene='TP53')
```



# Part 2. Designing a personalized cancer vaccine

The following sequences resulted from an NGS analysis of patient healthy and tumor tissue.

> Q1. What are the tumor specific amino-acids?

```{r}
seq.file <- "https://bioboot.github.io/bimm143_W20//class-material/lecture18_sequences.fa"

library(bio3d)

seqs <- read.fasta(seq.file)
aln <- seqaln(seqs)
aln
```


```{r}
ide <- conserv(aln$ali, method="identity")
mutant.sites <- which(ide < 1)
mutant.sites
```


We are including the gap positions here and this is not what we want.
So we need to exclude the gap positions

```{r}
## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
#gaps$f.inds


mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```

Find the -8 before and +8 after (i.e. the sequence segment containing all possible 9-mer sequences with a tumor specific amino-acid)

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8
```

```{r}
aln$ali[2,start.position[1]:end.position[1] ]
```


```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```



```{r}
# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```


```{r}
write.fasta(seqs=store.seqs,
            ids = mutant.names,
            file="mutantseqs.fa", gap=FALSE)
```
