---
title: "class 16: Pathway Analysis from RNA-seq."
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import RNA-Seq counts and metadata

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

Here we read the two input files 


```{r}
# Import metadata and take a peak
colData = read.csv("GSE37704_metadata.csv", row.names=1)
head(colData)

countData =read.csv("GSE37704_featurecounts.csv", row.names=1)
head(countData)
```
 
 There are 'r nrow(countData)' genes in this dataset and 'r ncol(countData)' experiments (samples)...
 
 Look like there is a funny extra 'length' column in our couuntData that we need to remove
 
```{r}
#Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```
 
 A little bit more tidying work required. Lets remove the gene were we have no data (i.e. zero counts in all experiments)
 
Let's take the rowSum() and exclude the rows with zero values
!= not equals
```{r}
countData <- countData[rowSums(countData) != 0,] 
```

Check how many gene we have remaining
```{r}
nrow(countData)
```

## Running DESeq2

now we get to run DESeq on our data
```{r}
library(DESeq2)

dds <- DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds <- DESeq(dds)
```

Let's see our results

```{r}
res <- results(dds)
res
```

## Volcono plot
Summary volcono plot of our results to date.

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(h=-log(0.05), col="red")
abline(v=c(-2,2), col="red")
```
Lets add some color...

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```


# Adding gene annotation
Since we mapped and counted against the Ensembl annotation, our results only have information about Ensembl gene IDs. However, our pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. So lets add them as we did the last day.

I would install these with
`BiocManager::install("AnnotationDbi")` and
`BiocManager::install("org.Hs.eg.db")`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

# What is availabvle in the org.Hs.eg.db dataset?
columns(org.Hs.eg.db)
```

Lets add extra annotation to our result object
```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
head(res)
```
Now we have our geneIDs in the correct format we can do pathway anaysis

##Section 2. Pathway Analysis
Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

Frist I need to install the packages I need 

```{r}
#BiocManager::install( c("pathview", "gage", "gageData") )
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Make a input vector of our numeric quantity of interest i.e. fold change values and make sure it has ENTREZ gene names associated with it
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now run **gage** with this input vector and our KEGG genesets

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)

```

What is in our results? We can use the `attributes()` function to see what is returned here.

```{r}
attributes(keggres)
```


```{r}
# Look at the first few down (less) pathways
head(keggres$less)

```

We use the KEGG pathway ID from our top hit along with our 'foldchange' vector to view our results as a more useful figure 
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![hsg04110.pathview.png]()

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

```{r}
knitr::include_graphics("hsa00140.pathview.png")
```

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

```{r}
knitr::include_graphics("hsa00140.pathview.png")
```

##Section 3. Gene Ontology (GO)
We can also do a similar procedure with gene ontology. Similar to above, go.sets.hs has all GO terms. go.subs.hs is a named list containing indexes for the BP, CC, and MF ontologies. Letâ€™s focus on BP (a.k.a Biological Process) here.
```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)

```

